Restructure docs

matevz opened the pull request:
Fixes #184, #125, #203, #10.

New features:
- adds remark plugin which converts github.com documentation links from various Oasis repos to relative docs links,
- implements correct `Edit page` link across all components and repositories,
- implements cross-component doc cards
- adds guidelines for writing Oasis docs (docs structure, cross-referencing, using doccards etc.)
- moves redirects to separate file
- bumps Docusaurus to v2.1.0

Structural changes:
- splits documentation into `general`, `get-involved`, `node`, `dapp`, `paratime`, and `core`. Now the chapters are not repository-based anymore (e.g. docs of oasis-sdk git module now became some sections of `dapp` chapter).
- all path changes are backward-compatible by adding redirects.
- moves aforementioned chapter links to the top bar. Support and website links were moved to the top-right corner icon and the footer respectively,
- removes usage of multiple `plugin-content-docs` plugins by unifying the site structure and taking `docs/` as a site root
- moves Mainnet and Testnet categories from `general` to `node` and Prerequisites, Maintenance, Advanced sections to `run-your-node` category
- moves General->Set up your node to Run Node->Run your Node; moves `genesis-doc.md` from `general/manage-tokens` to `node`
- adds `dapp` with Emerald and Sapphire tutorials and oasis-sdk/docs/contract for Cipher section
- adds `paratime` chapter containing oasis-sdk/docs/runtime files
- renames all `index.md`, `overview.md`, `getting-started.md` files to `README.md`. Always open it directly when clicking on the category link. Use generated table of contents with `generated-index` doc page type, if no feasible information exists.
- moves `token-delivery-and-kyc` to `get-involved` chapter and the remaining `general/community-resources` to `get-involved`.
- moves General->FAQ to General->Oasis network->FAQ

matevz committed:
docs: Use globe icon for website, populate footer with support and yt links

matevz committed:
docs: Reorganize sidebars

matevz committed:
docs: Migrate /general/run-a-node -> /operators

matevz committed:
docs: Migrated /general/developer-resources -> /developers

kostko commented:
Nice!

Probably "Oasis Core" should go under "Developers" instead of being its own tab?

matevz committed:
docs: Migrated adrs

matevz committed:
docs: Revamp introductory chapters, cross-component references

- sidebarUtils.js: Rename findSidebarItem() -> findGeneralSidebarItem()
- sidebarUtils.js: Add newItem() helper

matevz committed:
docs/general: Flatten "*/overview" chapters with category

matevz committed:
docs: Migrate /oasis-core -> /core

kostko reviewed (commented):
On the [Developers -> Oasis SDK](https://deploy-preview-200--trusting-archimedes-14c863.netlify.app/developers/sdk/) page it seems that the "Oasis Core Runtime Layer" link points to GitHub.

matevz commented:
> On the [Developers -> Oasis SDK](https://deploy-preview-200--trusting-archimedes-14c863.netlify.app/developers/sdk/) page it seems that the "Oasis Core Runtime Layer" link points to GitHub.

I was thinking of solving those in a separate PR https://github.com/oasisprotocol/docs/issues/203.

matevz committed:
src/remark: Convert github.com documentation links to local relative links

matevz commented:
> On the [Developers -> Oasis SDK](https://deploy-preview-200--trusting-archimedes-14c863.netlify.app/developers/sdk/) page it seems that the "Oasis Core Runtime Layer" link points to GitHub.

Now implemented as part of this PR.

matevz committed:
src: Implement edit page link across different repositories and components

gw0 commented:
Because restructuring the docs is a big deal we should not rush it. In general I like the new structure, but I still have a few suggestions.

- Mainnet Network Parameters and Mainnet Upgrade Logs should probably go into a Mainnet category under Node Operators
- Testnet Network Parameters and Testnet Upgrade Logs should probably go into a Testnet category under Node Operators
- Developers / Oasis SDK / Smart Contracts should probably be named Cipher ParaTime (similar to Emerald and Sapphire) and contain the API reference
- Developers / Oasis SDK / Runtimes should probably have a more informative/call-to-action name, like Your Custom ParaTime or Roll your own ParaTime
- I suspect we have three different groups of developers: contracts, runtimes, and core (last two could be merged?). Maybe we should rename "Developers" to "Contract Developers" and "Core" into "Core Developers"  (and also include Oasis SDK / Runtimes here).

kostko commented:
> Developers / Oasis SDK / Smart Contracts should probably be named Cipher ParaTime (similar to Emerald and Sapphire) and contain the API reference

On the other hand this seems weird as the Contract SDK documentation is general for any paratime using the `contracts` module.

kostko commented:
In case we want to make this Cipher-specific we should probably move these contract docs outside the SDK repository? But then we are left without general contract SDK docs not specific to any paratime deployment?

kostko commented:
> I suspect we have three different groups of developers: contracts, runtimes, and core (last two could be merged?). Maybe we should rename "Developers" to "Contract Developers" and "Core" into "Core Developers" (and also include Oasis SDK / Runtimes here).

Agree that the SDK docs should probably be separate, especially if we move out the "Cipher ParaTime" docs? So the SDK contains a bunch of things (as mentioned in the docs):
* Runtime SDK (for building runtimes)
* Client SDK (for building clients that talk to runtimes)
* WASM Contracts SDK (for building smart contracts using the WASM contrats module)

Note that all of these are general and unrelated to any specific runtime/paratime deployment.

But saying "Core Developers" and "Contract Developers" is way too long. In my mind the categories are something like:
* contracts (evm / confidential evm / wasm),
* SDK and
* Core.

So maybe we should move the general WASM contracts section (and related API reference links) out under Contracts -> Cipher if this is possible to do.

kostko commented:
Also one thing that is strange is the "Network Primer" (the "Why Oasis" and "ROSE Token Metrics" links lead there) as it seems like it is a completely separate thing instead of being a section in the main docs. I think it should be merged into the main docs now that we have reduced clutter there.

kostko commented:
> * Mainnet Network Parameters and Mainnet Upgrade Logs should probably go into a Mainnet category under Node Operators
> * Testnet Network Parameters and Testnet Upgrade Logs should probably go into a Testnet category under Node Operators

Yeah this is probably right. Same with the delegation policy.

gw0 commented:
> Note that all of these are general and unrelated to any specific runtime/paratime deployment.
> 
> In my mind the categories are something like:
>     * contracts (evm / confidential evm / wasm),
>     * SDK and
>     * Core.

Makes sense. Initially it has been bugging me that we promote ParaTimes, but then the docs aren't organized around them. But I am beginning to think that it makes a lot of sense to keep node operator and all developer docs paratime-agnostic and focused on the technological/functional aspects.

*Lets imagine the following user story: I heard about the confidentiality features (replace with any other feature) of Sapphire (replace with other ParaTime). As a developer I want to develop a smart contract on Sapphire. Where do I start?*

Because we promote ParaTimes, the solution would be if each ParaTime has a dedicated category under "General". It should clearly describe what it is (from [contribute-to-the-network/run-a-paratime-node](https://deploy-preview-200--trusting-archimedes-14c863.netlify.app/general/contribute-to-the-network/run-a-paratime-node)) and which technologies are being used with links to relevant developer docs.

matevz committed:
docs: Migrate /oasis-network-primer -> /general/oasis-network subpages

matevz committed:
docs: Moved Mainnet and Testnet to /operators

- also moved delegation-policy.md to contribute-to-the-network/
- moved network-paramters.md to operators/mainnet

matevz committed:
docs: Move operators/{prerequisites|maintenance-guides|advanced} to operators/set-up-your-node

- replace generic {prerequisites|maintenance-guides|advanced}/README.md files with generated-index
- move operators/troubleshooting.md to operators/set-up-your-node

matevz commented:
The mainnet and testnet categories are now moved under `operators/`. I also moved prerequisites, maintenance-guides and advanced under set-up-your-node to reduce top-level clutter. Please check.

kostko commented:
Looks like the "Network Parameters" are still in the General sidebar and clicking on it switches to the Node Operators tab.

matevz committed:
docs: Rename general/{oasis-network|manage-tokens}/overview.mdx -> README.mdx

matevz commented:
> Looks like the "Network Parameters" are still in the General sidebar and clicking on it switches to the Node Operators tab.

FYI this is deliberate so you can access current network parameters with a single click.

kostko commented:
Why would someone that is not a node operator need to access network parameters?

matevz commented:
> Why would someone that is not a node operator need to access network parameters?

Fair point. :thinking: I was considering metamask wallet users, but URLs for emerald/sapphire endpoints are now in the Developers component and it takes the same amount of clicks to come there via network parameters page. I agree, let's remove the Network parameters sidebar entry.

gw0 commented:
> I was considering metamask wallet users, but URLs for emerald/sapphire endpoints are now in the Developers component

Because Metamask wallet users are not developers, I would suggest that every ParaTime has a dedicated page in the "General" component and the Contract/SDK/Core developer components focus on the technologies (mentioned in https://github.com/oasisprotocol/docs/pull/200#issuecomment-1225840574).

lukaw3d commented on docs/developers/emerald-paratime/writing-dapps-on-emerald.md:
```diff
@@ -34,10 +34,10 @@ account to Emerald. Consult the [How to transfer ROSE into Emerald ParaTime]
 
 
-[overview]: ../../oasis-network/overview.md
+[overview]: /general/oasis-network/
```
do we care that this breaks links when viewing in github?

matevz replied:
Currently, `../../oasis-network` would point outside of the *developers* component and docusaurus won't let you compile this. We could remove the separate @docusaurus/plugin-content-docs per component and take docs/ folder as the root for everything. I would need to think about it, if that breaks anything else.

FYI referencing markdown files policy is defined here: https://github.com/oasisprotocol/docs/blob/matevz/feature/restructure-docs/README.md#referencing-documents-between-the-components

matevz replied:
Ok, I "unified" all content plugins now so the cross-component relative links to .md files should work.

lukaw3d commented on docusaurus.config.js:
```diff
@@ -136,25 +239,67 @@ const config = {
+            items: [
+              {
+                label: 'Discord',
+                href: 'https://discord.gg/RwNTK8t',
```
This is the old broken link

lukaw3d replied:
After all discord invites are pJdWeVtmHT we can validate if invite has expired by e.g.

- adding to package.json
```json
  "blc": "linkinator http://localhost:3000/general https://discord.com/api/v9/invites/pJdWeVtmHT",
```

- or adding to markdown somewhere
```md
[](https://discord.com/api/v9/invites/pJdWeVtmHT)
```

matevz replied:
Thanks, fixed.

lukaw3d commented on docs/general/oasis-network/why-oasis.md:
```diff
@@ -97,58 +91,35 @@ The Oasis Network is a Layer 1, proof-of-stake, decentralized network. It has tw
 The Oasis Network’s impressive scalability is achieved through a cutting-edge set of features that provide faster transaction speeds and higher throughput than other networks. The top-tier performance of the network is largely due to its separation of compute and consensus operations into the Consensus Layer and ParaTime Layer. This separation allows multiple ParaTimes to process transactions in  ...
 
-![Scalability](images/paratimes.png)
+![Scalability](../images/architecture/technology_scalability.svg)
```
Note: this has different keywords in the image

| Before | After |
| --- | --- |
| ![](https://raw.githubusercontent.com/oasisprotocol/docs/f99925c354db18f914410eb453604523b771a05a/docs/oasis-network-primer/images/paratimes.png) | ![](https://raw.githubusercontent.com/oasisprotocol/docs/f99925c354db18f914410eb453604523b771a05a/docs/general/images/architecture/technology_scalability.svg) |

matevz replied:
Yeah, that was obsolete.

matevz committed:
docs: Unify plugin-content-docs into a single tree

amela commented:
My suggestions for node operators:
- [x] We create a subcategory `Run a node` on the `Prerequisites` level and put in this category Validator node, Non-Validator node, Seed node, etc. I think `Run a ...` is too repetitive.
- [x] Rename `Sentry Node Architecture` to `Sentry Node` and move it to `Run a node` category.
- [x] Move `Amend Commission Schedule` to `Maintenance` 
- [ ] Rename `Mainnet` and `Testnet` to `Mainnet Parameters` and `Testnet Parameters`
- [x] I think `Create an Entity Package` is deprecated.

matevz committed:
src: Unify sidebarUtils item lookup

matevz committed:
submodules: Bump all submodules to latest main/master branches

matevz committed:
docs/operators: Refactored set-up-your-node category

matevz committed:
developers: Revamp Oasis SDK and ParaTimes guides

- Also rename how-to-transfer-rose-into-evm-paratime -> how-to-transfer-rose-into-paratime

matevz committed:
deps: Bump Docusaurus to 2.1.0

- removes docs/general/manage-tokens/holding-rose-tokens/README.md in favor of autogenerated-index

matevz committed:
general: Remove redundant community-resources pages

matevz committed:
core: Revert development-setup subcategories

matevz commented:
> > I suspect we have three different groups of developers: contracts, runtimes, and core (last two could be merged?). Maybe we should rename "Developers" to "Contract Developers" and "Core" into "Core Developers" (and also include Oasis SDK / Runtimes here).
> 
> Agree that the SDK docs should probably be separate, especially if we move out the "Cipher ParaTime" docs? So the SDK contains a bunch of things (as mentioned in the docs):
> 
>     * Runtime SDK (for building runtimes)
> 
>     * Client SDK (for building clients that talk to runtimes)
> 
>     * WASM Contracts SDK (for building smart contracts using the WASM contrats module)
> 
> 
> Note that all of these are general and unrelated to any specific runtime/paratime deployment.
> 
> But saying "Core Developers" and "Contract Developers" is way too long. In my mind the categories are something like:
> 
>     * contracts (evm / confidential evm / wasm),
> 
>     * SDK and
> 
>     * Core.
> 
> 
> So maybe we should move the general WASM contracts section (and related API reference links) out under Contracts -> Cipher if this is possible to do.

I thought a bit about this and refactored the Developers component now. Please check: https://deploy-preview-200--trusting-archimedes-14c863.netlify.app/developers/

kostko commented:
There seems to be some missing spacing:
![image](https://user-images.githubusercontent.com/543739/188487940-47b61dd8-9bd5-4c97-8e48-83f2c5349f7f.png)

Otherwise the new "Developers -> Ciper ParaTime" section is better and we can improve the SDK even more. The "Developers -> Oasis SDK -> Client" section seems a bit empty for now, probably we should move it to the bottom (as it doesn't provide anything beyond a link to the reference) instead of being at the top?

matevz commented:
> There seems to be some missing spacing: ![image](https://user-images.githubusercontent.com/543739/188487940-47b61dd8-9bd5-4c97-8e48-83f2c5349f7f.png)

Fixed.

> Otherwise the new "Developers -> Ciper ParaTime" section is better and we can improve the SDK even more. The "Developers -> Oasis SDK -> Client" section seems a bit empty for now, probably we should move it to the bottom (as it doesn't provide anything beyond a link to the reference) instead of being at the top?

Let's use the same order as in the SDK's README then:
- runtime
- contract
- client

matevz committed:
config: Move redirects to separate file

tjanez commented:
Thanks for this great work @matevz!

My main comment before we merge this restructuring would be around the naming and structuring of different sub-documentations.

Current variant:
- General
- Node Operators
- Developers
- Core
sounds inconsistent and a bit boring to me.

I'm thinking how we can make it more attractive and sort of "calling to action".

One idea would be:
- Use Oasis (I don't know what URL to suggest. Ideas?)
  - I would move/duplicate the user-relevant part of how to use Emerald and Cipher on Mainnet here.
- Run Node (or Operate Node) (URL: `/node`)
- Create dApp (URL: `/dapp`)
  - I would change the structure to list the 3 ParaTimes with subsections on how to develop dApps for them.
- Build ParaTime (URL: `/paratime`)
  - Things currently under Oasis SDK -> Runtimes
- Develop Core (URL: `/core` or `/oasis-core`. Preferences?)

![Screenshot from 2022-09-08 15-15-14](https://user-images.githubusercontent.com/2518544/189131503-fdda9658-167f-4dc2-a98f-87a929bec2e7.png)



We could add this as well:
- Join Community (or Contribute) (URL: `/community` or `/contribute`) that would include:
  - Some general info about our community and how to get in touch. similar to https://docs.avax.network/community.
  - Things we currently have under General -> Contribute to the Network.

matevz committed:
docs: Major refactor of chapters

tjanez commented:
I think this looks very good.

I would try to improve/augment the "Join Community" and generalize it to "Get Involved" with sections:
"Join our Community" -> page with links to our communities
"Become our Ambassador" -> external link to https://oasisprotocol.org/ambassador-program
"Become our Influencer" -> external link to: https://oasisprotocol.org/influencer
"Apply for Grant" -> https://oasisprotocol.org/grant-programs
"Join our University Program" -> https://oasisprotocol.org/university-program
"Run a Node" with subsections:
  - "Consensus Validator Node"
  - "ParaTime Node"

"Develop Oasis Core" (rename current "Contributing Guidelines")
"Network Governance"
"Delegation Policy"
"Token Delivery & KYC Guide"

tjanez reviewed (commented):
Thanks again, @matevz!
